snapshot/snapshotter.go

```diff
@@ -0,0 +1,347 @@
+package snapshot
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/docker/containerd"
+	"github.com/docker/containerd/testutil"
+	"github.com/stretchr/testify/assert"
+)
+
+// Kind identifies the kind of snapshot.
+type Kind int
+
+// definitions of snapshot kinds
+const (
+	KindActive Kind = iota
+	KindCommitted
+)
+
+// Info provides information about a particular snapshot.
+type Info struct {
+	Name     string // name or key of snapshot
+	Parent   string // name of parent snapshot
+	Kind     Kind   // active or committed snapshot
+	Readonly bool   // true if readonly, only valid for active
+}
+
+// Snapshotter defines the methods required to implement a snapshot snapshotter for
+// allocating, snapshotting and mounting filesystem changesets. The model works
+// by building up sets of changes with parent-child relationships.
+//
+// A snapshot represents a filesystem state. Every snapshot has a parent, where
+// the empty parent is represented by the empty string. A diff can be taken
+// between a parent and its snapshot to generate a classic layer.
+//
+// An active snapshot is created by calling `Prepare`. After mounting, changes
+// can be made to the snapshot. The act of commiting creates a committed
+// snapshot. The committed snapshot will get the parent of active snapshot. The
+// committed snapshot can then be used as a parent. Active snapshots can never
+// act as a parent.
+//
+// Snapshots are best understood by their lifecycle. Active snapshots are
+// always created with Prepare or View. Committed snapshots are always created
+// with Commit.  Active snapshots never become committed snapshots and vice
```

Comment:
To clarify: when you're writing to an active snapshot and wish to have it committed, you call `Commit` which creates a new `KindCommitted` snapshot and leaves your existing `KindActive` snapshot untouched?